include/cuco/priority_queue.cuh

```diff
@@ -0,0 +1,238 @@
+   *        can be converted to T
+   * @param first Beginning of the sequence of elements
+   * @param last End of the sequence of elements
+   * @param stream The stream in which the underlying device operations will be
+   *               executed
+   */
+  template <typename InputIt>
+  void push(InputIt first, InputIt last, cudaStream_t stream = 0);
+
+  /**
+   * @brief Remove a sequence of the lowest elements ordered by Compare
+   *
+   * @tparam OutputIt Device accessible output iterator whose `value_type`
+   *        can be converted to T
+   * @param first Beginning of the sequence of output elements
+   * @param last End of the sequence of output elements
+   * @param stream The stream in which the underlying GPU operations will be
+   *               run
+   */
+  template <typename OutputIt>
+  void pop(OutputIt first, OutputIt last, cudaStream_t stream = 0);
+
+  /*
+  * @brief Return the amount of shared memory required for operations
+  *        on the queue with a thread block size of block_size
+  *
+  * @param block_size Size of the blocks to calculate storage for
+  * @return The amount of temporary storage required in bytes
+  */
+  int get_shmem_size(int block_size) {
+    int intersection_bytes = 2 * (block_size + 1) * sizeof(int);
+    int node_bytes = node_size_ * sizeof(T);
+    return intersection_bytes + 2 * node_bytes;
+  }
+
+  /**
+   * @brief Destroys the queue and frees its contents
+   */
+  ~priority_queue();
+
+  class device_mutable_view {
+   public:
+
+    /**
+     * @brief Push elements into the priority queue
+     *
+     * @tparam CG Cooperative Group type
+     * @tparam InputIt Device accessible iterator whose `value_type`
+     *         is convertible to T
+     * @param g The cooperative group that will perform the operation
+     * @param first The beginning of the sequence of elements to insert
+     * @param last The end of the sequence of elements to insert
+     * @param temp_storage Pointer to a contiguous section of memory
+     *        large enough to hold get_shmem_size(g.size()) bytes
+     */
+    template <typename CG, typename InputIt>
+    __device__ void push(CG const& g, InputIt first,
+                         InputIt last, void *temp_storage);
+
+    /**
+     * @brief Pop elements from the priority queue
+     *
+     * @tparam CG Cooperative Group type
+     * @tparam OutputIt Device accessible iterator whose `value_type`
+     *         is convertible to T
+     * @param g The cooperative group that will perform the operation
+     * @param first The beginning of the sequence of elements to output into
+     * @param last The end of the sequence of elements to output into
+     * @param temp_storage Pointer to a contiguous section of memory
+     *        large enough to hold get_shmem_size(g.size()) bytes
+     */
+    template <typename CG, typename OutputIt>
+    __device__ void pop(CG const& g, OutputIt first,
+                        OutputIt last, void *temp_storage);
```

Comment:
Is it possible to make the temp storage be passed at construction of the view instead of at the operation? That would be more canonical with a CUB like pattern, e.g., BlockReduce:
```
    // Specialize BlockReduce for a 1D block of 128 threads on type int
    using BlockReduce =  cub::BlockReduce<int, 128>;
    // Allocate shared memory for BlockReduce
    __shared__ typename BlockReduce::TempStorage temp_storage;
    // Obtain a segment of consecutive items that are blocked across threads
    int thread_data[4];
    ...
    // Compute the block-wide sum for thread0
    int aggregate = BlockReduce(temp_storage).Sum(thread_data);
```

Reply:
Currently, the device view is constructed on the host side, as is done in `cuco::static_map`. One possibility would be to have a function that allows the user to set the temporary memory that the device view uses. 

Maybe another option would be to have an initial, immutable device view that is constructed on the host side, and then passed to a kernel. Once in device code, the user could construct a mutable device view from the immutable device view by passing the temp storage.

Does one of those approaches sound good?